src/unix/encoding.ml

```diff
@@ -0,0 +1,201 @@
+  let refill buf =
+    let len = min (String.length str - !p) De.io_buffer_size in
+    Bigstringaf.blit_from_string str ~src_off:!p buf ~dst_off:0 ~len;
+    p := !p + len;
+    len
+  in
+  let flush buf len =
+    let str = Bigstringaf.substring buf ~off:0 ~len in
+    Buffer.add_string r str
+  in
+  Gz.Higher.compress ~w ~q ~level ~refill ~flush () cfg i o;
+  Buffer.contents r
+
+let inflate_string ~algorithm str =
+  match algorithm with
+  | `Deflate -> inflate_string_de str
+  | `Gzip -> inflate_string_gz str
+
+let deflate_string ~algorithm str =
+  match algorithm with
+  | `Deflate -> deflate_string_de str
+  | `Gzip -> deflate_string_gz str
+
+let encoding_of_string = function
+  | "deflate" -> `Deflate
+  | "gzip" -> `Gzip
+  | s -> `Unknown s
+
+let content_encodings request =
+  match Message.header request "content-encoding" with
+  | None -> None
+  | Some s ->
+    String.split_on_char ',' s
+    |> List.map (fun x -> x |> String.trim |> String.lowercase_ascii)
+    |> List.map encoding_of_string
+    |> Option.some
+
+let accepted_encodings_with_weights request =
+  match Message.header request "accept-encoding" with
+  | None -> None
+  | Some s ->
+    let encodings = Accept.encodings (Some s) |> Accept.qsort in
+    Some
+      (List.map
+         (fun (a, b) ->
+           ( (match b with
+             | Accept.Any -> `Any
+             | Accept.Gzip -> `Gzip
+             | Accept.Compress -> `Compress
+             | Accept.Deflate -> `Deflate
+             | Accept.Identity -> `Identity
+             | Accept.Encoding s -> `Unknown s),
+             a ))
+         encodings)
+
+let accepted_encodings request =
+  match accepted_encodings_with_weights request with
+  | None -> None
+  | Some encodings -> Some (List.map (fun (a, _) -> a) encodings)
+
+let preferred_content_encoding request =
+  match accepted_encodings request with
+  | None -> None
+  | Some l ->
+    let rec aux = function
+      | [] -> None
+      | `Any :: _rest -> Some `Gzip
+      | `Deflate :: _rest -> Some `Deflate
+      | `Gzip :: _rest -> Some `Gzip
+      | _ :: rest -> aux rest
+    in
+    aux l
+
+let algorithm_to_string = function
+  | `Deflate -> "deflate"
+  | `Gzip -> "gzip"
+
+let with_encoded_body ?(algorithm = `Deflate) body response =
+  match body with
+  | "" -> response
+  | _ ->
+    let encoded_body = deflate_string ~algorithm body in
+    Message.set_body response encoded_body;
+    Message.set_header response "Content-Encoding"
+      (algorithm_to_string algorithm);
+    response
+
+let compress handler req =
+  let%lwt response = handler req in
+  let preferred_algorithm = preferred_content_encoding req in
+  match preferred_algorithm with
+  | None -> Lwt.return response
+  | Some algorithm ->
+    log.info (fun log ->
+        log ~request:req "Compressing request with algorithm: %s"
+          (algorithm_to_string algorithm));
+    let%lwt body = Message.body response in
```

Comment:
Note: it should be possible to have a streaming compressor/decompressor with the current Dream API. I will probably write an example that does a streaming transformation of a body stream in some way, and a compressor/decompressor would be built analogously.